plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtGetUnmanagedInstancesCommandWrapper.java

```diff
@@ -0,0 +1,219 @@
+import org.libvirt.Connect;
+import org.libvirt.Domain;
+import org.libvirt.LibvirtException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@ResourceWrapper(handles=GetUnmanagedInstancesCommand.class)
+public final class LibvirtGetUnmanagedInstancesCommandWrapper extends CommandWrapper<GetUnmanagedInstancesCommand, GetUnmanagedInstancesAnswer, LibvirtComputingResource> {
+    private static final Logger LOGGER = Logger.getLogger(LibvirtPrepareUnmanageVMInstanceCommandWrapper.class);
+
+    @Override
+    public GetUnmanagedInstancesAnswer execute(GetUnmanagedInstancesCommand command, LibvirtComputingResource libvirtComputingResource) {
+        LOGGER.info("Fetching unmanaged instance on host");
+
+        HashMap<String, UnmanagedInstanceTO> unmanagedInstances = new HashMap<>();
+        try {
+            final LibvirtUtilitiesHelper libvirtUtilitiesHelper = libvirtComputingResource.getLibvirtUtilitiesHelper();
+            final Connect conn = libvirtUtilitiesHelper.getConnection();
+            final List<Domain> domains = getDomains(command, libvirtComputingResource, conn);
+
+            for (Domain domain : domains) {
+                UnmanagedInstanceTO instance = getUnmanagedInstance(libvirtComputingResource, domain, conn);
+                unmanagedInstances.put(instance.getName(), instance);
+                domain.free();
+            }
+        } catch (Exception e) {
+            LOGGER.error("GetUnmanagedInstancesCommand failed due to " + e.getMessage());
+            throw new CloudRuntimeException("GetUnmanagedInstancesCommand failed due to " + e.getMessage());
+        }
+
+        return new GetUnmanagedInstancesAnswer(command, "True", unmanagedInstances);
+    }
+
+    private List<Domain> getDomains(GetUnmanagedInstancesCommand command,
+                                    LibvirtComputingResource libvirtComputingResource,
+                                    Connect conn) throws LibvirtException, CloudRuntimeException {
+        final List<Domain> domains = new ArrayList<>();
+        final String vmNameCmd = command.getInstanceName();
+        if (StringUtils.isNotBlank(vmNameCmd)) {
+            final Domain domain = libvirtComputingResource.getDomain(conn, vmNameCmd);
+            if (domain == null) {
+                LOGGER.error("GetUnmanagedInstancesCommand: vm not found " + vmNameCmd);
+                throw new CloudRuntimeException("GetUnmanagedInstancesCommand: vm not found " + vmNameCmd);
+            }
+
+            // Filter instance if answer is requested for a particular instance name
+            if (StringUtils.isNotEmpty(vmNameCmd) &&
+                    !vmNameCmd.equals(domain.getName())) {
+                LOGGER.error("GetUnmanagedInstancesCommand: exact vm name not found " + vmNameCmd);
+                throw new CloudRuntimeException("GetUnmanagedInstancesCommand: exact vm name not found " + vmNameCmd);
+            }
```

Comment:
this comment and block is suitable as a separate method

Reply:
done